server/core/src/main/scala/org/infinispan/server/core/transport/ExtendedByteBuf.scala

```diff
@@ -57,6 +59,151 @@ object ExtendedByteBuf {
+            if ((b & 0x80) == 0) Some(i)
+            else {
+               if (count > 9)
+                  throw new IllegalStateException(
+                     "Stream corrupted.  A variable length long cannot be longer than 9 bytes.")
+
+               if (buf.readableBytes() >= 1) {
+                  val bb = buf.readByte
+                  read(buf, bb, shift + 7, i | (bb & 0x7FL) << shift, count + 1)
+               } else {
+                  buf.resetReaderIndex()
+                  None
+               }
+            }
+         }
+         read(bf, b, 7, b & 0x7F, 1)
+      } else {
+         bf.resetReaderIndex()
+         None
+      }
+   }
+
+   /**
+    * Reads a variable size int if possible.  If not present the reader index is reset to the last mark.
+    * @param bf
+    * @return
+    */
+   def readMaybeVInt(bf: ByteBuf): Option[Int] = {
+      if (bf.readableBytes() >= 1) {
+         val b = bf.readByte
+         @tailrec def read(buf: ByteBuf, b: Byte, shift: Int, i: Int, count: Int): Option[Int] = {
+            if ((b & 0x80) == 0) Some(i)
+            else {
+               if (count > 5)
+                  throw new IllegalStateException(
+                     "Stream corrupted.  A variable length integer cannot be longer than 5 bytes.")
+
+               if (buf.readableBytes() >= 1) {
+                  val bb = buf.readByte
+                  read(buf, bb, shift + 7, i | ((bb & 0x7FL) << shift).toInt, count + 1)
+               } else {
+                  buf.resetReaderIndex()
+                  None
+               }
+            }
+         }
+         read(bf, b, 7, b & 0x7F, 1)
+      } else {
+         bf.resetReaderIndex()
+         None
+      }
+   }
+
+   /**
+    * Reads a range of bytes if possible.  If not present the reader index is reset to the last mark.
+    * @param bf
+    * @return
+    */
+   def readMaybeRangedBytes(bf: ByteBuf): Option[Array[Byte]] = {
+      val length = readMaybeVInt(bf)
+      if (length.isDefined) {
+         if (bf.readableBytes() >= length.get) {
+           if (length.get > 0) {
+             val array = new Array[Byte](length.get)
+             bf.readBytes(array)
+             Some(array)
+           } else {
+             Some(Array[Byte]())
+           }
+         } else {
+           bf.resetReaderIndex()
+           None
+         }
+      } else None
+   }
+
+  def readMaybeRangedBytes(bf: ByteBuf, length: Int): Option[Array[Byte]] = {
+     if (bf.readableBytes() < length) {
+       bf.resetReaderIndex()
+       None
+     } else {
+       val bytes = new Array[Byte](length)
+       bf.readBytes(bytes)
+       Some(bytes)
+     }
+  }
+
+  def readMaybeSignedInt(bf: ByteBuf): Option[Int] = {
+    readMaybeVInt(bf).map(i => SignedNumeric.decode(i))
+  }
+
+   def readMaybeOptRangedBytes(bf: ByteBuf): Option[Option[Array[Byte]]] = {
+      val l = readMaybeSignedInt(bf)
+      if (l.isDefined) {
```

Comment:
I agree they are the norm, the problem is the logic for the whole segment was quite a bit confusing imo and took quite a bit of tweaking to get it right.  It might just be because I wasn't used to scala as much though.